pubtools/pulplib/_impl/model/repository/repo_lock.py

```diff
@@ -0,0 +1,163 @@
+import json
+import random
+import string
+from time import sleep
+import os
+import logging
+
+from datetime import datetime, timedelta
+from more_executors.futures import f_map
+
+LOCK_CLAIM_STR = "pulplib-lock-claim-"
+
+LOCK_VALID_FROM_OFFSET = os.getenv("PULPLIB_LOCK_VALID_FROM_OFFSET", 10)
+LOCK_EXPIRATION_OFFSET = os.getenv("PULPLIB_LOCK_EXPIRATION_OFFSET", 1800)
+
+AWAIT_ACTIVATION_SLEEP = os.getenv("PULPLIB_AWAIT_ACTIVATION_SLEEP", 10)
+
+REPO_ENDPOINT_BASE = "pulp/api/v2/repositories/%s/"
+
+LOG = logging.getLogger("pubtools.pulplib")
+
+
+class LockClaim(object):
+    def __init__(self, expiration_offset=LOCK_EXPIRATION_OFFSET, **kwargs):
+        self.id = (
+            kwargs["id"]
+            if "id" in kwargs
+            else "".join(random.choices(string.ascii_letters + string.digits, k=8))
+        )
+        self.created = (
+            datetime.fromtimestamp(kwargs["created"])
+            if "created" in kwargs
+            else datetime.now()
+        )
+        self.valid_from = (
+            datetime.fromtimestamp(kwargs["valid_from"])
+            if "valid_from" in kwargs
+            else self.created + timedelta(seconds=LOCK_VALID_FROM_OFFSET)
+        )
+        self.expires = (
+            datetime.fromtimestamp(kwargs["expires"])
+            if "expires" in kwargs
+            else self.created + timedelta(seconds=expiration_offset)
+        )
+
+    @property
+    def as_json(self):
+        data = {
+            "id": self.id,
+            "created": datetime.timestamp(self.created),
+            "valid_from": datetime.timestamp(self.valid_from),
+            "expires": datetime.timestamp(self.expires),
+        }
+        return json.dumps(data)
+
+    def __lt__(self, other):
+        return self.created < other.created
+
+    def __eq__(self, other):
+        return self.id == other.id
+
+    @staticmethod
+    def from_json_data(json_data):
+        return LockClaim(**json_data)
+
+    @property
+    def is_valid(self):
+        return self.expires > datetime.now()
+
+
+class RepoLock(object):
+    def __init__(
+        self,
+        repo_id,
+        request_func,
+        expiration_offset=LOCK_EXPIRATION_OFFSET,
+        activation_await_sleep=AWAIT_ACTIVATION_SLEEP,
+    ):
+        self._repo_id = repo_id
+        self._request_func = request_func
+        self._lock_claim = None
+        self._expiration_offset = expiration_offset
+        self._activation_await_sleep = activation_await_sleep
+
+    def __enter__(self):
+        LOG.info("Attempting to lock %s." % self._repo_id)
```

Comment:
Can you please let the logging framework do the string formatting instead, like this:

```suggestion
        LOG.info("Attempting to lock %s.", self._repo_id)
```

This is actually a pylint complaint that we should usually obey - https://pylint.pycqa.org/en/latest/user_guide/messages/warning/logging-format-interpolation.html. Many legacy projects ignore it but we should try to stick with it in the newer stuff.